client/service/src/builder.rs

```diff
@@ -749,7 +749,8 @@ pub trait ServiceBuilderCommand {
 		self,
 		input: impl Read + Seek + Send + 'static,
 		force: bool,
-	) -> Pin<Box<dyn Future<Output = Result<(), Error>> + Send>>;
+		binary: bool,
+	) -> Result<Pin<Box<dyn Future<Output = Result<(), Error>> + Send>>, Error>;
```

Comment:
The future is already returning an error, you can don't need to wrap this again in a `Result`.

Reply:
I might be wrong but this sounds like something you can check earlier on. The builder is a bit "too late". If it's just to make sure that the `count` variable can be decoded, maybe you should add a parser that will fail in the cli before it reaches service.

If you do need to check it later in service, then returning a future that fails is good too and probably easier than having a result of a future of a result (sounds over-engineered but I might be better, I don't really know).

Reply:
> I might be wrong but this sounds like something you can check earlier on. The builder is a bit "too late". If it's just to make sure that the `count` variable can be decoded, maybe you should add a parser that will fail in the cli before it reaches service.
> 
> If you do need to check it later in service, then returning a future that fails is good too and probably easier than having a result of a future of a result (sounds over-engineered but it might be better, I don't really know).

Oh yeah tbh I had not thought about moving this logic before hand. I do see an issue with doing this though:
if I wish to try and decode a `count`, then I need to take the input and transform it to a `CodecIoReader` (provided by parity-scale-codec). But I can't work with an `CodecIoReader` if I'm doing JSON, so I'd basically need to move all the logic of the `new` function of `BlockStream` out of this file.

And, even without this issue, it still seems like we would be de-encapsulating the code: we'd effectively be moving some of that logic to another part of the code... which seems a bit confusing to me. Let me know if I'm missing something though! :)

Reply:
No no that's fine. Then fallback to point 2:

> If you do need to check it later in service, then returning a future that fails is good too and probably easier than having a result of a future of a result (sounds over-engineered but it might be better, I don't really know).

Do you think you can get rid of the `Result<Future<Result<_>>>` by failing inside the future instead of outside?

Reply:
Yup I'm editing this right now :)